signer/signer.go

```diff
@@ -0,0 +1,491 @@
+			lint.EtsiEsi,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	var lk crypto.Signer
+	switch k := config.Issuer.PublicKey.(type) {
+	case *rsa.PublicKey:
+		lk, err = rsa.GenerateKey(rand.Reader, k.Size()*8)
+		if err != nil {
+			return nil, err
+		}
+		profile.sigAlg = x509.SHA256WithRSA
+	case *ecdsa.PublicKey:
+		lk, err = ecdsa.GenerateKey(k.Curve, rand.Reader)
+		if err != nil {
+			return nil, err
+		}
+		switch k.Curve {
+		case elliptic.P256():
+			profile.sigAlg = x509.ECDSAWithSHA256
+		case elliptic.P384():
+			profile.sigAlg = x509.ECDSAWithSHA384
+		default:
+			return nil, fmt.Errorf("unsupported ECDSA curve: %s", k.Curve.Params().Name)
+		}
+	default:
+		return nil, errors.New("unsupported issuer key type")
+	}
+	s := &Signer{
+		issuer:  config.Issuer,
+		signer:  config.Signer,
+		clk:     config.Clk,
+		lints:   lints,
+		lintKey: lk,
+		profile: profile,
+	}
+	return s, nil
+}
+
+var ctPoisonExt = pkix.Extension{
+	// OID for CT poison, RFC 6962 (was never assigned a proper id-pe- name)
+	Id:       asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 11129, 2, 4, 3},
+	Value:    asn1.NullBytes,
+	Critical: true,
+}
+
+// OID for SCT list, RFC 6962 (was never assigned a proper id-pe- name)
+var sctListOID = asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 11129, 2, 4, 2}
+
+func generateSCTListExt(scts []ct.SignedCertificateTimestamp) (pkix.Extension, error) {
+	list := ctx509.SignedCertificateTimestampList{}
+	for _, sct := range scts {
+		sctBytes, err := cttls.Marshal(sct)
+		if err != nil {
+			return pkix.Extension{}, err
+		}
+		list.SCTList = append(list.SCTList, ctx509.SerializedSCT{Val: sctBytes})
+	}
+	listBytes, err := cttls.Marshal(list)
+	if err != nil {
+		return pkix.Extension{}, err
+	}
+	extBytes, err := asn1.Marshal(listBytes)
+	if err != nil {
+		return pkix.Extension{}, err
+	}
+	return pkix.Extension{
+		Id:    sctListOID,
+		Value: extBytes,
+	}, nil
+}
+
+var mustStapleExt = pkix.Extension{
+	// RFC 7633: id-pe-tlsfeature OBJECT IDENTIFIER ::=  { id-pe 24 }
+	Id: asn1.ObjectIdentifier{1, 3, 6, 1, 5, 5, 7, 1, 24},
+	// ASN.1 encoding of:
+	// SEQUENCE
+	//   INTEGER 5
+	// where "5" is the status_request feature (RFC 6066)
+	Value: []byte{0x30, 0x03, 0x02, 0x01, 0x05},
+}
+
+func generateSKID(pk crypto.PublicKey) ([]byte, error) {
+	pkBytes, err := x509.MarshalPKIXPublicKey(pk)
+	if err != nil {
+		return nil, err
+	}
+	var pkixPublicKey struct {
+		Algo      pkix.AlgorithmIdentifier
+		BitString asn1.BitString
+	}
+	if _, err := asn1.Unmarshal(pkBytes, &pkixPublicKey); err != nil {
+		return nil, err
+	}
+	skid := sha1.Sum(pkixPublicKey.BitString.Bytes)
+	return skid[:], nil
+}
+
+// Issue generates a certificate from the provided issuance request and
+// signs it. Before the signing the certificate with the issuers private
+// key it is signed using a throwaway key so that it can be linted using
+// zlint. If the linting fails an error is returned and the certificate
+// is not signed using the issuers key.
```

Comment:
```suggestion
// Issue generates a certificate from the provided issuance request and
// signs it. Before signing the certificate with the issuer's private
// key, it is signed using a throwaway key so that it can be linted using
// zlint. If the linting fails, an error is returned and the certificate
// is not signed using the issuer's key.
```